### PR TITLE
Further enhancements to nameof

### PR DESCRIPTION
**Pull request by Tarmil**
Feature suggestion: https://github.com/fsharp/fslang-suggestions/issues/841

* [x] Implementation
* [x] Add tests
* [x] Check syntax coloring
* [x] Check code completion

### CODE REVIEW

**Comment by CartBlanche** [comment]
@Tarmil Will you be updating this?

**Comment by Tarmil** [comment]
Yeah I'll get back to this soon, hopefully the conflict shouldn't be too hard to resolve.

**Comment by dsyme** [comment]
I've merged master into this.  I'll likely integrate the changes into a final PR for nameof

File: tests/fsharp/core/nameof/preview/test.fsx [hunk 1]
```diff
@@ -294,6 +294,30 @@ type Person =
...
+module RecTest = 
+    let rec [<Literal>] two = 2
+    and twoName = nameof(two)
+    let ``can get name of recursive literal`` () =
+        Assert.AreEqual("two", twoName)
+
+module rec RecTest2 = 
+    let [<Literal>] two = 2
+    let twoName = nameof(two)
+    let ``can get name of recursive literal`` () =
```
Comment (cartermp) [inline tests/fsharp/core/nameof/preview/test.fsx:318]:
```suggestion
    let ``can get name of literal in recursive module`` () =
```

**Comment by dsyme** [comment]
This is now ready - it now contains all the design adjustments for nameof

* [x] `nameof` in patterns
* [x] `nameof` for operators returns the source text
* [x] Further testing for nameof on generic parameters

File: tests/fsharp/core/nameof/preview/test.fsx [hunk 2]
```diff
@@ -294,6 +300,30 @@ type Person =
         | x when x = nameof __.Age -> { __ with Age = value :?> int }
         | _ -> __
 
+type GenericClassNameOfTests<'TTT>() =
+
+    static member ``can get name of class type parameter`` () =
+        let b = nameof<'TTT>
+        Assert.AreEqual("TTT", b)
```
Comment (cartermp) [inline tests/fsharp/core/nameof/preview/test.fsx:307]: What about something like this?

```fsharp
type C<'TTT> = class end

nameof<'TTT>
```

Or is that nonsense?

Comment (auduchinok) [reply tests/fsharp/core/nameof/preview/test.fsx:307]: `'TTT` is not in scope outside the type declaration here, I wouldn't expect it to work.

Comment (cartermp) [reply tests/fsharp/core/nameof/preview/test.fsx:307]: Yeah, that's true. Makes sense not to add it.

**Comment by KevinRansom** [comment]
Error on neg16.fs:

````

 + neg16.fs(116,9,116,13): typecheck error FS0879: Volatile fields must be marked 'mutable' and cannot be thread-static
diff between [D:\a\1\s\tests\fsharp\typecheck\sigs\neg16.bsl] and [D:\a\1\s\tests\fsharp\typecheck\sigs\neg16.vserr]
line 92
 - neg16.fs(113,9,113,11): typecheck error FS0879: Volatile fields must be marked 'mutable' and cannot be thread-static
 + neg16.fs(130,10,130,11): typecheck error FS0935: Types with the 'AllowNullLiteral' attribute may only inherit from or implement types which also allow the use of the null literal
diff between [D:\a\1\s\tests\fsharp\typecheck\sigs\neg16.bsl] and [D:\a\1\s\tests\fsharp\typecheck\sigs\neg16.vserr]
diff at line 92
- 
- neg16.fs(116,9,116,13): typecheck error FS0879: Volatile fields must be marked 'mutable' and cannot be thread-static
- 
- neg16.fs(130,10,130,11): typecheck error FS0935: Types with the 'AllowNullLiteral' attribute may only inherit from or implement types which also allow the use of the null literal
"
  Stack Trace:
     at Microsoft.FSharp.Core.PrintfModule.PrintFormatToStringThenFail@1639.Invoke(String message) in D:\a\1\s\src\fsharp\FSharp.Core\printf.fs:line 1639
   at SingleTest.singleVersionedNegTest@451-41.Invoke(String arg60) in D:\a\1\s\tests\fsharp\single-test.fs:line 451
   at SingleTest.singleVersionedNegTest(TestConfig cfg, String version, String testname) in D:\a\1\s\tests\fsharp\single-test.fs:line 451
   at SingleTest.singleNegTest(TestConfig cfg, String testname) in D:\a\1\s\tests\fsharp\single-test.fs:line 453
   at FSharp.Tests.Core.TypecheckTests.type check neg16() in D:\a\1\s\tests\fsharp\tests.fs:line 2399
  Standard Output Messages:
 ------------------ D:\a\1\s\tests\fsharp\typecheck\sigs ---------------

````

**Comment by dsyme** [comment]
@KevinRansom @cartermp this is ready

**Comment by dsyme** [comment]
> This is missing test cases for patterns - at a minimum, the example from fsharp/fslang-suggestions#841 should be used

thanks, will add those now

**Comment by abelbraaksma** [comment]
From the original post by @Tarmil here I noticed two unchecked items that I didn't see in the commits, are these tracked/added elsewhere or already part of this PR? Namely, syntax coloring and code completion.

image

**Comment by abelbraaksma** [comment]
Also, there are a bunch of `nameof` related bug reports. If it helps, I can checkout this PR to see which are resolved, and link back to the issue (limited to bugs, not new feature requests) where applicable, and/or suggest to add tests: https://github.com/dotnet/fsharp/search?q=nameof&type=Issues

**Comment by cartermp** [comment]
@dsyme @Tarmil I don't have access to the fork but this is the test that should probably be added;

```fsharp
module PatternMatchingWithNameof =
    /// Simplified version of EventStore's API
    type RecordedEvent = { EventType: string; Data: string }

    /// My concrete type:
    type MyEvent =
        | A of string
        | B of string
    let ``Can use nameof as a pattern in pattern match expression`` () =
        let deserialize (e: RecordedEvent) : MyEvent =
            match e.EventType with
            | nameof A -> A e.Data
            | nameof B -> B e.Data
            | t -> failwithf "Invalid EventType: %s" t

        let getData event =
            match event with
            | A amsg -> amsg
            | B bmsg -> bmsg

        let re1 = { EventType = nameof A; Data = "hello" }
        let re2 = { EventType = nameof B; Data = "world" }

        let a = deserialize re1
        let b = deserialize re2

        Assert.AreEqual(getData a, re1.Data)
        Assert.AreEqual(getData b, re2.Data)
```

**Comment by dsyme** [comment]
> @dsyme @Tarmil I don't have access to the fork but this is the test that should probably be added;

Thank you, I've added this test and got it workign (pattern matchin had indeed regressed)

File: tests/fsharp/core/nameof/preview/test.fsx [hunk 3]
```diff
@@ -343,6 +373,41 @@ do test "lookup name of a generic class"                ((NameOfOperatorForGener
...
+        | B bmsg -> bmsg
+
+    let re1 = { EventType = nameof A; Data = "hello" }
+    let re2 = { EventType = nameof B; Data = "world" }
+
+    let a = deserialize re1
+    let b = deserialize re2
+
+    check "fklwveoihwq1" (getData a) re1.Data
+    check "fklwveoihwq2" (getData b) re2.Data
```
Comment (cartermp) [inline tests/fsharp/core/nameof/preview/test.fsx:408]: I'm not familiar with this bespoke(ish) system for testing. Do these get executed?

**Comment by cartermp** [comment]
@abelbraaksma we're still in the active development cycle for F# 5 and so we won't block this on any existing bugs unless it's something obvious like the test gap I noticed. We're aware of the current set and will take another look once we have everything we want in.

File: tests/fsharp/core/nameof/preview/test.fsx (hunk 3 above)
Comment (dsyme) [reply tests/fsharp/core/nameof/preview/test.fsx:408]: Yup, it's used throughout tests\fsharp.

**Comment by dsyme** [comment]
> @abelbraaksma we're still in the active development cycle for F# 5 and so we won't block this on any existing bugs unless it's something obvious like the test gap I noticed. We're aware of the current set and will take another look once we have everything we want in.

I'm fixing all the outstanding bugs right now, just finalising it

**Comment by dsyme** [comment]
@cartermp The latest changes fix #8661 and #7416

I'd like to add some symbol resolution tests as well to the FCS service tests to help check #8661.  However that can be done separately (the tests are currently as PITA to iterate on until they are moved into FSharp.Compiler.UnitTests, currently part of #8907 )

**Comment by dsyme** [comment]
Had a weird package download failure

**Comment by abelbraaksma** [comment]
> Had a weird package download failure

I've also noticed them from time to time, and can also happen locally (they popup as a special error in VS). Usually, they disappear on next build. 

OT: I was wondering if these package references could be made more resilient by allowing a fixed number of retries. Just a thought ;)

**Comment by cartermp** [comment]
fsharpqa failures, our favorite ;)

**Comment by dsyme** [comment]
OK, the updates I just pushed should fix the fsharpqa failures hopefully

**Comment by cartermp** [comment]
https://youtu.be/AJQ3TM-p2QI

**Comment by cartermp** [comment]
Same failure - I feel like it should either be deleted or moved to non-FSharpQA since there's no diagnostics on why it failed

**Comment by dsyme** [comment]
@cartermp I agree

Tbh I've confirmed the case by hand, I think we can deal with this separately. 

So as far as I'm concerned this is now ready

File: src/fsharp/TypeChecker.fs [hunk 4]
```diff
@@ -9278,38 +9293,70 @@ and TcNameOfExpr cenv env tpenv (synArg: SynExpr) =
...
-            let overallTy = match overallTyOpt with None -> NewInferenceType() | Some t -> t 
-            
-            // This will raise an error if resolution doesn't succeed
-            let _, _ = TcLongIdentThen cenv overallTy env tpenv lidd delayed
-            result // checked as an expression, done with checks
+            // If it's not an expression or type name then resolve it as a module
+            let resolvedToModuleOrNamespaceName =
+                if delayed.IsEmpty then
+                    let id,rest = List.headAndTail longId
+                    match ResolveLongIndentAsModuleOrNamespaceOrStaticClass cenv.tcSink ResultCollectionSettings.AllResults cenv.amap m false true OpenQualified env.eNameResEnv ad id rest true with
```
Comment (cartermp) [inline src/fsharp/TypeChecker.fs:9348]: Note that with open type declarations coming in this call may no longer do what is assumed, see here: https://github.com/dotnet/fsharp/pull/9513/files#diff-dbb9e59b8ac1e1e04e1b00282a91ec97R2192

Does that have any impact? I would expect not...

**Comment by TIHan** [comment]
I'm going to be looking at this today.

**Comment by KevinRansom** [comment]
> Instead of having `nameof` be a special case in name resolution, couldn't we just add an `Obsolete` attribute with this message?:

nameof has existed in FSharp.Core for over a year, so deprecating it will not reliably work. :-(

**Comment by TIHan** [comment]
I need to add tests cases for enums now.

**Comment by TIHan** [comment]
> nameof has existed in FSharp.Core for over a year, so deprecating it will not reliably work. :-(

`nameof` is preview and its implementation throws an exception as it's treated specially by the compiler. It's fine that we put an Obsolete attribute on it.

**Comment by dsyme** [comment]
@KevinRansom shouldn't `nameof` have this?

            [<Experimental(ExperimentalAttributeMessages.RequiresPreview)>]

**Comment by cartermp** [comment]
The experimental attribute is only while things are in preview. Since we are moving nameof to F# 5 (which is still in preview, so no guarantees about current behavior as-shipped yet), that attribute was removed. We're also removing it for most other things now that we're nearing the end of the development cycle.

**Comment by cartermp** [comment]
The current experience today is that you can still call `nameof` in non-F# 5 code, but it's a compile error. So if Obsolete makes that a bit clearer, that's great. If it filters `nameof` out of a completion list for older F# users, then that's also great.

**Comment by dsyme** [comment]
I am trialling this, it should do the trick for giving a reasonable message to old compilers https://github.com/dotnet/fsharp/pull/8754/commits/a25541a3c566f5c738d51953bf99be17c3df72dd

**Comment by dsyme** [comment]
> I am trialling this, it should do the trick for giving a reasonable message to old compilers a25541a

If this doesn't give the result we need then we can forget it I think

**Comment by dsyme** [comment]
THis is ready to go

**Comment by KevinRansom** [comment]
@dsyme , attributes on nameof are not going to be particularly helpful, because the method has already shipped in fsharp.core.dll since I think 4.7.0.

**Comment by cartermp** [comment]
I'm fine with the outcome of this. `nameof` was only ever supported as preview and subject to change. Anyone who is bitten by this will simply have to make a change due to the preview nature of the feature.

I think in practice people aren't going to really be hit by this though.

**Comment by dsyme** [comment]
I think adding the attribute is best regarded as a good bug fix to a preview feature

**Comment by KevinRansom** [comment]
Certainly there is no harm doing so.

**Comment by dsyme** [comment]
Let's merge this baby and enable this in 5.0
